code/game/machinery/firealarm.dm

```diff
@@ -148,6 +148,9 @@
 		alarm(user)
 
 /obj/machinery/firealarm/attack_ai(mob/user)
+	if(!ai_can_use(ENVIROMENTAL_CONTROL, user))
+		to_chat(user, "<span class='warning'>No Enviromental Control relay detected. Unable to interface with fire alarm.</span>")
```

Comment:
Having to rewrite this every single time is extremely inefficient and wasteful. Move this to "ai_can_use()"

Reply:
Gah then we have to encode the name of the relay board somehow, got any nifty ideas?

Reply:
#define ENVIRONMENTAL_CONTROL "environmental control relay"

Could work, then just use that string as your deciding factor for what relay means what. 

GLOBAL_LIST_EMPTY(relays)

relays["environmental_control"] += src or whatever on the relay

Perhaps?

Reply:
(then on can_use)

if(GLOB.relays[control_type]?.on)
//Success
else
to_chat(user, "You need a [control_type] relay, online, to do that.")

Reply:
They’re bitflags so you can make dual relays.

Easiest and ugliest way would be to just hardcode some text

Reply:
Ooh fancy.

Hmm, then this becomes a bit harder. You could potentially keep the above idea but separate it by bitflag:

if(control_types & RELAY_DOORS)
GLOB.relays[RELAY_DOORS] += src

?

Then:

You need a [GLOB.relays[RELAY_DOORS]][1].name I suppose, but that sounds shitcodey the more I think about it 😕

Reply:
Can we get the value of a define if we have the string of the define name?

Reply:
DEFINE, to my knowledge, is just text replacement. The compiler changes all your instances of FOO into 1 if you #define FOO 1

Reply:
Yeah that’s what I thought

Reply:
Worst case we make a list with the bit flag values as keys and the string as the value? Can we do that?

Reply:
That doesn't sound too terrible tbh. You could also just lazycode it and go:

(can_ai_use)
if(interaction_type & ENVIRONMENTAL_CONTROL) to_chat(user, "you need EV control tbh")

Reply:
Yeah that’s the ultra lazy way